### ISSUE TITLE
Incorrect unreachable error

### ISSUE THREAD

**Issue by snmishra** [python/mypy#16718]
**Bug Report**

Tests with `re.match` or `re.search` give incorrect unreachable error when using mypy 1.8.0

**To Reproduce**

```python
import re

test_str = "XYZ"

if re.match("ABC", test_str):
    print("match")
else:
    print("no match")
```

**Expected Behavior**

No mypy errors

**Actual Behavior**
```
> mypy test.py
test.py:10: error: Statement is unreachable  [unreachable]
        print("no match")
        ^~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.8.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
```
python_version = "3.12"
check_untyped_defs = true
# We don't have stubs for all libraries
ignore_missing_imports = true
# We need some ignores for Pylance
follow_imports = "normal"
ignore_errors = false
show_error_codes = true
pretty = true
# Strictness
allow_redefinition = false
implicit_reexport = true
local_partial_types = true
no_implicit_optional = true
strict_equality = true
strict_optional = false
# Warnings
warn_no_return = true
warn_redundant_casts = true
warn_unreachable = true
warn_unused_configs = true
warn_unused_ignores = true
```
- Python version used: 3.12

**Comment by hauntsaninja** [comment]
This is expected given that you're using `strict_optional = false`. Don't use it, it is evil

**Comment by snmishra** [comment]
Thanks! `strict_optional = False` indeed is evil and doing a lot more than I had expected from the name. The documentation probably should have a fat warning that this is evil.

**Comment by HansAarneLiblik** [comment]
I don't think this should be closed. What change is the reason behind this new failure?

For us the failing code is strictly typed

```py
def func(inp: str) -> str | None:
  if re.match(MY_REGEX, inp):
    return inp
    
  return None  # This errors as "unreachable"
```

We have a pretty big codebase, where A LOT of code is still using `strict_optional = false`, as the migration to `strict_optional=true` takes time.

**Comment by hauntsaninja** [comment]
Oh you didn't mention this was a regression. Looks like the relevant change is #16566, so mypy correctly sees that Match is final and cannot have a falsey subclass.

I don't think there's any change to be made here. `strict_optional = false` tells mypy to not think about None, so mypy will not think about None.

If this is a problem for your existing codebase, I recommend using `strict_optional = false` only in the modules that need it, and disabling `warn_unrachable = true` in those modules. This can be accomplished in your mypy config file or by using comments at the top of the file.

**Comment by HansAarneLiblik** [comment]
Match may be final, but `re.match` returns `Match[str] | None` and we're relying on the `| None` part

We're already using `strict_optional=false` ONLY on manually defined list of pre-existing files (migrating to `true` is in the works), but as I said this takes time

![image](https://github.com/python/mypy/assets/5418540/7da5f473-8cde-4e15-9f37-2991a98859c2)

We can ignore this error on the impacted functions, but IMO this is a regression not a bug in our "outdated" codebase
